pkg/soc/soc.go

```diff
@@ -0,0 +1,243 @@
+// of the payload.
+func NewUpdate(id Id, payload []byte) *Update {
+	return &Update{
+		id:      id,
+		payload: payload,
+		span:    int64(len(payload)),
+	}
+}
+
+// WithSpan is a chainable function that adds an arbitrary
+// span to the chunk data defined for the update.
+func (s *Update) WithSpan(span int64) *Update {
+	s.span = span
+	return s
+}
+
+// WithOwnerAddress provides the possibility of setting the ethereum
+// address for the owner of an update in the absence of a signer.
+func (s *Update) WithOwnerAddress(ownerAddress *Owner) *Update {
+	s.owner = ownerAddress
+	return s
+}
+
+// AddSigner currently sets a single signer for the soc update.
+//
+// This method will overwrite any value set with WithOwnerAddress with
+// the address derived from the given signer.
+func (s *Update) AddSigner(signer crypto.Signer) error {
+	publicKey, err := signer.PublicKey()
+	if err != nil {
+		return err
+	}
+	ownerAddressBytes, err := crypto.NewEthereumAddress(*publicKey)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "%x", ownerAddressBytes)
+	ownerAddress, err := NewOwner(ownerAddressBytes)
+	if err != nil {
+		return err
+	}
+	s.signer = signer
+	s.owner = ownerAddress
+	return nil
+}
+
+// OwnerAddress returns the ethereum address of the signer of the Update.
+func (s *Update) OwnerAddress() []byte {
+	return s.owner.address
+}
+
+// Address returns the soc Chunk address of the update.
+func (s *Update) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
+}
+
+// UpdateFromChunk recreates an Update from swarm.Chunk data.
+func UpdateFromChunk(ch swarm.Chunk) (*Update, error) {
+	chunkData := ch.Data()
+	if len(chunkData) < minUpdateSize {
+		return nil, errors.New("less than minimum length")
+	}
+
+	// add all the data fields
+	update := &Update{}
+	cursor := 0
+
+	update.id = chunkData[cursor : cursor+IdSize]
+	cursor += IdSize
+
+	update.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
+
+	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
+	span := binary.LittleEndian.Uint64(spanBytes)
+	update.span = int64(span)
+	cursor += swarm.SpanSize
+
+	update.payload = chunkData[cursor:]
+
+	bmtPool := bmtlegacy.NewTreePool(swarm.NewHasher, swarm.Branches, bmtlegacy.PoolSize)
+	bmtHasher := bmtlegacy.New(bmtPool)
+
+	// calculate the bmt hash of the update payload
+	err := bmtHasher.SetSpan(int64(span))
+	if err != nil {
+		return nil, err
+	}
+	_, err = bmtHasher.Write(update.payload)
+	if err != nil {
+		return nil, err
+	}
+	payloadSum := bmtHasher.Sum(nil)
+
+	// recover owner information
+	recoveredPublicKey, err := crypto.Recover(update.signature, payloadSum)
```

Comment:
it is not the payloadSum that is signed. but the payloadSum against the id

Reply:
in that case I would encourage revising the figure in the book.